lens/lotus/lotus.go

```diff
@@ -6,60 +6,72 @@ import (
 	"net/http"
 	"strings"
 
-	"github.com/filecoin-project/sentinel-visor/lens"
-
+	"github.com/filecoin-project/go-jsonrpc"
+	lotus_api "github.com/filecoin-project/lotus/api"
+	"github.com/filecoin-project/lotus/api/client"
+	lcli "github.com/filecoin-project/lotus/cli"
+	lru "github.com/hashicorp/golang-lru"
 	logging "github.com/ipfs/go-log/v2"
 	ma "github.com/multiformats/go-multiaddr"
 	manet "github.com/multiformats/go-multiaddr/net"
 	"github.com/urfave/cli/v2"
 	"golang.org/x/xerrors"
 
-	"github.com/filecoin-project/go-jsonrpc"
-	lotus_api "github.com/filecoin-project/lotus/api"
-	"github.com/filecoin-project/lotus/api/client"
-	lcli "github.com/filecoin-project/lotus/cli"
+	"github.com/filecoin-project/sentinel-visor/lens"
 )
 
 var log = logging.Logger("visor/lens/lotus")
 
-func GetFullNodeAPI(cctx *cli.Context) (context.Context, lens.API, lens.APICloser, error) {
+type APIOpener struct {
+	// TODO: replace dependency on cli.Context with tokenMaddr and repo path
+	cctx  *cli.Context
+	cache *lru.ARCCache // cache shared across all instances of the api
+}
+
+func NewAPIOpener(cctx *cli.Context, cacheSize int) (*APIOpener, lens.APICloser, error) {
+	ac, err := lru.NewARC(cacheSize)
+	if err != nil {
+		return nil, nil, xerrors.Errorf("new arc cache: %w", err)
+	}
+
+	o := &APIOpener{
+		cctx:  cctx,
+		cache: ac,
+	}
+
+	return o, lens.APICloser(func() {}), nil
+}
+
+func (o *APIOpener) Open(ctx context.Context) (lens.API, lens.APICloser, error) {
 	var api lotus_api.FullNode
 	var closer jsonrpc.ClientCloser
 	var err error
 
-	if tokenMaddr := cctx.String("api"); tokenMaddr != "" {
+	if tokenMaddr := o.cctx.String("api"); tokenMaddr != "" {
 		toks := strings.Split(tokenMaddr, ":")
 		if len(toks) != 2 {
-			return nil, nil, nil, fmt.Errorf("invalid api tokens, expected <token>:<maddr>, got: %s", tokenMaddr)
+			return nil, nil, fmt.Errorf("invalid api tokens, expected <token>:<maddr>, got: %s", tokenMaddr)
 		}
 
-		api, closer, err = getFullNodeAPIUsingCredentials(cctx.Context, toks[1], toks[0])
+		api, closer, err = getFullNodeAPIUsingCredentials(ctx, toks[1], toks[0])
 		if err != nil {
-			return nil, nil, nil, xerrors.Errorf("get full node api with credentials: %w", err)
+			return nil, nil, xerrors.Errorf("get full node api with credentials: %w", err)
 		}
 	} else {
-		api, closer, err = lcli.GetFullNodeAPI(cctx)
+		api, closer, err = lcli.GetFullNodeAPI(o.cctx)
```

Comment:
why when we have token credentials do we use the `ctx` passed in to `Open` as the lifetime of the API, versus when we don't (and look on disk for the api endpoint) we use the context passed to create the APIOpener instead?

Reply:
As mentioned above they are different types. As far as I can tell GetFullNodeAPI expects the cli context so it can read the flags and environment.

Reply:
Perhaps @frrist or @mg could explain further since this code comes from the original spike

Reply:
but in the first case you could provide `o.cctx.Context` for consistency though

Reply:
We should always prefer the context.Context passed to a function because it may have tracing or metrics keys in it. o.cctx.Context would be the original context derived from context.Background by the CLI package.

Reply:
then lets change `GetFullNodeAPI` take in an explicit context so that one can also use the context passed in to the opener?

What i'm worried about is that the decision to pass in credentials as a flag shouldn't change the lifetime of the API as it's unrelated. that seems like it can easily lead to a fairly counter-intuitive bug to track down at some point in the future.

Reply:
Agree this is a wart. 

GetFullNodeAPI is in lotus. Can we decouple this change and that one?

Reply:
ah. i see the issue now 😢 

we could do something like:
```
o.cctx.Context = ctx
api, closer, err = lcli.GetFullNodeAPI(o.cctx)
```

maybe would want to save/restore the full context to make sure it doesn't get lost.

Reply:
Can you expand on "the decision to pass in credentials as a flag shouldn't change the lifetime of the API"  - I don't think I understand why it would change it.

Reply:
(since the lifetime of the API is always the duration of a single run of the task that is using it)

Reply:
i guess i meant more "which context is used" - if an extra tag (or a cancel context) is added to the context passed to the `Open` method, it will be applied if there is a `--api ` flag and the upper case of instantiation is used, but not if there isn't and only the original app context goes to the API.

Having consistency of the context used for the API node / json RPC seems desirable

Reply:
> Perhaps frrist or mg could explain further since this code comes from the original spike

Not 100% sure on the question, but it looks like the methods called interanlly by `lcli.GetFullNodeAPI` are all exported meaning `lcli.GetFullNodeAPI` could be re-implemented here for simpler context handling.